extensions/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/KafkaIndexTask.java

```diff
@@ -0,0 +1,584 @@
+              stillReading = !assignment.isEmpty();
+            }
+          }
+        }
+      }
+
+      // Persist pending data.
+      final Committer finalCommitter = committerSupplier.get();
+      driver.persist(finalCommitter);
+
+      publishing = true;
+      if (stopping) {
+        // Stopped gracefully. Exit code shouldn't matter, so fail to be on the safe side.
+        return TaskStatus.failure(getId());
+      }
+
+      final TransactionalSegmentPublisher publisher = new TransactionalSegmentPublisher()
+      {
+        @Override
+        public boolean publishSegments(Set<DataSegment> segments, Object commitMetadata) throws IOException
+        {
+          // Sanity check, we should only be publishing things that match our desired end state.
+          if (!ioConfig.getEndPartitions().equals(((Map) commitMetadata).get(METADATA_NEXT_PARTITIONS))) {
+            throw new ISE("WTF?! Driver attempted to publish invalid metadata[%s].", commitMetadata);
+          }
+
+          final SegmentInsertAction action;
+
+          if (ioConfig.isUseTransaction()) {
+            action = new SegmentInsertAction(
+                segments,
+                new KafkaDataSourceMetadata(ioConfig.getStartPartitions()),
+                new KafkaDataSourceMetadata(ioConfig.getEndPartitions())
+            );
+          } else {
+            action = new SegmentInsertAction(segments, null, null);
+          }
+
+          log.info("Publishing with isTransaction[%s].", ioConfig.isUseTransaction());
+
+          return toolbox.getTaskActionClient().submit(action).isSuccess();
+        }
+      };
+
+      final SegmentsAndMetadata published = driver.finish(publisher, committerSupplier.get());
+      if (published == null) {
+        throw new ISE("Transaction failure publishing segments, aborting");
+      } else {
+        log.info(
+            "Published segments[%s] with metadata[%s].",
+            Joiner.on(", ").join(
+                Iterables.transform(
+                    published.getSegments(),
+                    new Function<DataSegment, String>()
+                    {
+                      @Override
+                      public String apply(DataSegment input)
+                      {
+                        return input.getIdentifier();
+                      }
+                    }
+                )
+            ),
+            published.getCommitMetadata()
+        );
+      }
+    }
+
+    return success();
+  }
+
+  @Override
+  public boolean canRestore()
+  {
+    return true;
+  }
+
+  @Override
+  public void stopGracefully()
+  {
+    log.info("Stopping gracefully.");
+
+    stopping = true;
+    if (publishing && runThread.isAlive()) {
+      log.info("stopGracefully: Run thread started publishing, interrupting it.");
+      runThread.interrupt();
+    }
+  }
+
+  @Override
+  public <T> QueryRunner<T> getQueryRunner(Query<T> query)
+  {
+    if (appenderator == null) {
+      return null;
```

Comment:
should we return NoopQueryRunner here, returning null here might fail the druid query, whereas its just that there is no data to respond at this point from this task.